sklearn/cluster/tests/test_hierarchical.py

```diff
@@ -600,6 +616,105 @@ def increment(self, *args, **kwargs):
     assert_equal(fa.counter, 3)
 
 
+def test_agglomerative_clustering_with_distance_threshold():
+    # Check that we obtain the correct number of clusters with
+    # agglomerative clustering with distance_threshold.
+
+    rng = np.random.RandomState(0)
+    mask = np.ones([10, 10], dtype=np.bool)
+    n_samples = 100
+    X = rng.randn(n_samples, 50)
+    connectivity = grid_to_graph(*mask.shape)
+    # test when distance threshold is set to 10
+    distance_threshold = 10
+    for linkage in ("ward", "complete", "average"):
+        for conn in [None, connectivity]:
+            clustering = AgglomerativeClustering(
+                n_clusters=None,
+                distance_threshold=distance_threshold,
+                connectivity=conn, linkage=linkage)
+            clustering.fit(X)
+            clusters_produced = clustering.labels_
+            num_clusters_produced = len(np.unique(clustering.labels_))
+            # test if the clusters produced match the point in the linkage tree
+            # where the distance exceeds the threshold
+            tree_builder = _TREE_BUILDERS[linkage]
+            children, n_components, n_leaves, parent, distances = \
+                tree_builder(X, connectivity=conn, n_clusters=None,
+                             return_distance=True)
+            num_clusters_at_threshold = np.count_nonzero(
+                distances >= distance_threshold) + 1
+            # test number of clusters produced
+            assert num_clusters_at_threshold == num_clusters_produced
+            # test clusters produced
+            clusters_at_threshold = _hc_cut(n_clusters=num_clusters_produced,
+                                            children=children,
+                                            n_leaves=n_leaves)
+            assert np.array_equiv(clusters_produced,
+                                  clusters_at_threshold)
+
+    rng = np.random.RandomState(0)
+    n_samples = 10
+    X = rng.randint(-3, 3, size=(n_samples, 3))
+    # this should result in all data in their own clusters
+    clustering = AgglomerativeClustering(
+        n_clusters=None,
+        distance_threshold=1,
+        linkage="single").fit(X)
+    assert len(np.unique(clustering.labels_)) == 10
+
+    # check the distances within the clusters and with other clusters
+    threshold = 2
+    clustering = AgglomerativeClustering(
+        n_clusters=None,
+        distance_threshold=threshold,
+        linkage="single").fit(X)
+    labels = clustering.labels_
+    D = pairwise_distances(X, metric="euclidean")
+    # to avoid taking the 0 diagonal in min()
+    np.fill_diagonal(D, np.inf)
+    for i in np.unique(labels):
+        in_cluster_mask = labels == i
+        max_in_cluster_distance = (D[in_cluster_mask][:, in_cluster_mask]
```

Comment:
`max_in_cluster_distance`  is always `inf`. 

related: every sample is in its own cluster, I'm not sure that was the intention of this test.

Reply:
nice catch!